Font fallback should ignore generic families for codepoints in PUA

vitorroriz opened the pull request:
#### 4535075b56fdb33a076f51fe92cd0bb8832797c8
<pre>
Font fallback should ignore generic families for codepoints in PUA
<a href="https://bugs.webkit.org/show_bug.cgi?id=263261">https://bugs.webkit.org/show_bug.cgi?id=263261</a>
rdar://115901340

Reviewed by Cameron McCormack.

According to spec: <a href="https://drafts.csswg.org/css-fonts-4/#char-handling-issues">https://drafts.csswg.org/css-fonts-4/#char-handling-issues</a>

&quot;If a given character is a Private-Use Area Unicode codepoint, user agents must only match font families named in the font-family list that are not generic families. If none of the families named in the font-family list contain a glyph for that codepoint, user agents must display some form of missing glyph symbol for that character rather than attempting installed font fallback for that codepoint.&quot;

We are currently not ignoring generic font families for font fallback when a code point is in the private-use area (PUA).
This patch changes that. Now FontRanges has a flag to signal that the Font represented by the FontRanges
object came from a generic family. That way, we can skip it during font fallback when finding
the glyph data for a codepoint that is in the private-user area.

After attempting all user-specified font-families, if we couldn&apos;t find a font that can represent such codepoint,
we then use the .notdef glyph (glyph 0) and the last resource font of WebKit for it.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA.html: Added.
* LayoutTests/platform/mac/fast/text/softbank-emoji-expected.txt:
* LayoutTests/platform/wpe/fast/text/softbank-emoji-expected.txt:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::realizeNextFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
* Source/WebCore/platform/graphics/FontRanges.cpp:
(WebCore::FontRanges::FontRanges):
(WebCore::FontRanges::glyphDataForCharacter const):
* Source/WebCore/platform/graphics/FontRanges.h:
(WebCore::FontRanges::isGeneric const):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/text/CharacterProperties.h:
(WebCore::isPrivateUseAreaCharacter):

Canonical link: <a href="https://commits.webkit.org/269524@main">https://commits.webkit.org/269524@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8f3f691eb838d213f1689084d5455b0765f71d17

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22816 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23904 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24725 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23073 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23343 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22011 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23056 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/301 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25578 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20667 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/26874 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/20661 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20913 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24731 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18172 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/280 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20470 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5435 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/445 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/411 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nedley commented on Source/WebCore/platform/text/CharacterProperties.h:
```diff
@@ -98,4 +98,9 @@ inline bool isControlCharacter(UChar32 character)
 
+inline bool isPrivateUseAreaCharacter(UChar32 character)
+{
+    return ublock_getCode(character) == UBlockCode::UBLOCK_PRIVATE_USE_AREA;
```
Since there are multiple private use blocks, this would be better off doing `(UCharCategory)u_charType(character) == U_PRIVATE_USE_CHAR`.

vitorroriz replied:
Thank you for the review!
I’m trying to understand what is the difference of a character whose block is `UBlockCode::UBLOCK_PRIVATE_USE_AREA` and a character whose general category is `U_PRIVATE_USE_CHAR`.

I thought the former would mean that the block which contains the unicode is one of: BMP, Plane15 or Plane16, which seems the same as a character that belongs to the general category "private use".

nedley replied:
A block is a range of characters, with `UBLOCK_PRIVATE_USE_AREA` being U+E000..FEFF. You raise a good issue, though, that it is possible for a platform to change the general category of private use characters (and I believe the system version of ICU does). I will double-check, but until then this function can also check for `UBLOCK_SUPPLEMENTARY_PRIVATE_USE_AREA_A` and `UBLOCK_SUPPLEMENTARY_PRIVATE_USE_AREA_B`.

vitorroriz replied:
ah now I get it, `UBLOCK_PRIVATE_USE_AREA` is just accounting for one of the blocks indeed:

>      * Until Unicode 3.1.1, the corresponding block name was "Private Use",
>      * and multiple code point ranges had this block.
>      * Unicode 3.2 renames the block for the BMP PUA to "Private Use Area" and
>      * adds separate blocks for the supplementary PUAs.

heycam commented on Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
```diff
@@ -421,7 +421,10 @@ const Font* FontCascade::fontForCombiningCharacterSequence(StringView stringView
 
     for (unsigned i = 0; !fallbackRangesAt(i).isNull(); ++i) {
-        const Font* font = fallbackRangesAt(i).fontForCharacter(baseCharacter);
+        auto fontRange = fallbackRangesAt(i);
```
`auto&` (otherwise it will make a copy of the `FontRanges`)

And since the type is `FontRanges`, should we call the vairable `fontRanges` instead of `fontRange`?

heycam commented on Source/WebCore/platform/graphics/FontRanges.h:
```diff
@@ -89,9 +90,12 @@ class FontRanges {
     WEBCORE_EXPORT const Font& fontForFirstRange() const;
     bool isLoading() const;
+    bool isGeneric() const { return m_isGeneric; }
+    void setIsGeneric(bool value) { m_isGeneric = value; }
```
This function is unused.

heycam commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -382,11 +382,14 @@ FontRanges CSSFontSelector::fontRangesForFamily(const FontDescription& fontDescr
 
     AtomString familyForLookup = familyName;
+    bool isGeneric = false;
     std::optional<FontDescription> overrideFontDescription;
```
While you're here, can you remove this unused variable?

heycam commented on Source/WebCore/css/CSSFontSelector.cpp:
```diff
@@ -399,7 +402,7 @@ FontRanges CSSFontSelector::fontRangesForFamily(const FontDescription& fontDescr
         if (document && document->settings().webAPIStatisticsEnabled())
             ResourceLoadObserver::shared().logFontLoad(*document, familyForLookup.string(), true);
-        return face->fontRanges(*fontDescriptionForLookup, fontPaletteValues, fontFeatureValues);
+        return { face->fontRanges(*fontDescriptionForLookup, fontPaletteValues, fontFeatureValues), isGeneric };
```
I slightly don't like that it's possible for `fontRanges()` to construct a `FontRanges` object here with an incorrect `m_isGeneric` value, and how we then create a new `FontRanges` with the correct value. I was going to suggest passing `isGeneric` into `CSSSegmentedFontFace::fontRanges`, so it could then construct the `FontRanges` with correct `m_isGeneric` value from the start, but looking at how `fontRanges()` maintains a cache of ranges (and how the "is generic" boolean would then probably have to be part of the cache key) makes me second guess it.

Which is to say, I'm OK with this.

heycam commented on Source/WebCore/platform/graphics/cocoa/FontFamilySpecificationCoreText.h:
```diff
@@ -43,6 +43,7 @@ class FontFamilySpecificationCoreText {
 
 private:
     RetainPtr<CTFontDescriptorRef> m_fontDescriptor;
+    bool isGeneric;
```
`m_isGeneric`

But is this used?

heycam commented on Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
```diff
@@ -518,7 +523,7 @@ GlyphData FontCascadeFonts::glyphDataForCharacter(UChar32 c, const FontCascadeDe
 
     GlyphData glyphData = cacheEntry.glyphDataForCharacter(c);
-    if (!glyphData.glyph) {
+    if (!glyphData.font) {
```
I think we'll always have a non-null font if we found a glyph (even if per your changes the glyph ID we chose is 0). And so I was going to suggest using `glyphData.isValid()`, for clarity, but reading that function makes me wonder if there are cases where the glyph ID is non-zero and the font is null, and that's a valid state for a `GlyphData` to have?

vitorroriz replied:
That's a good question. I don't know if this would be a valid state. My interpretation was that if the font or the glyph are not zero, then the other will also be non-zero. However we can have the case where we explicitly set the glyph to be the .notdef glyph (glyph id = 0) but this is still a valid state.

heycam commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA-ref.html:
```diff
@@ -0,0 +1,24 @@
+<!DOCTYPE html>
```
I don't think we want to check in a `-ref.html` file in additional to the `-expected.html`.

nt1m replied:
We do, this is pretty common with WPT reftests

heycam reviewed (approved):
r=me with Ned's comments addressed.

vitorroriz committed:
Font fallback should ignore generic families for codepoints in PUA
https://bugs.webkit.org/show_bug.cgi?id=263261
rdar://115901340

Reviewed by Cameron McCormack.

According to spec: https://drafts.csswg.org/css-fonts-4/#char-handling-issues

"If a given character is a Private-Use Area Unicode codepoint, user agents must only match font families named in the font-family list that are not generic families. If none of the families named in the font-family list contain a glyph for that codepoint, user agents must display some form of missing glyph symbol for that character rather than attempting installed font fallback for that codepoint."

We are currently not ignoring generic font families for font fallback when a code point is in the private-use area (PUA).
This patch changes that. Now FontRanges has a flag to signal that the Font represented by the FontRanges
object came from a generic family. That way, we can skip it during font fallback when finding
the glyph data for a codepoint that is in the private-user area.

After attempting all user-specified font-families, if we couldn't find a font that can represent such codepoint,
we then use the .notdef glyph (glyph 0) and the last resource font of WebKit for it.

* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/matching/font-unicode-PUA.html: Added.
* LayoutTests/platform/mac/fast/text/softbank-emoji-expected.txt:
* LayoutTests/platform/wpe/fast/text/softbank-emoji-expected.txt:
* Source/WebCore/css/CSSFontSelector.cpp:
(WebCore::CSSFontSelector::fontRangesForFamily):
* Source/WebCore/platform/graphics/FontCascadeFonts.cpp:
(WebCore::realizeNextFallback):
(WebCore::FontCascadeFonts::glyphDataForVariant):
(WebCore::FontCascadeFonts::glyphDataForCharacter):
* Source/WebCore/platform/graphics/FontRanges.cpp:
(WebCore::FontRanges::FontRanges):
(WebCore::FontRanges::glyphDataForCharacter const):
* Source/WebCore/platform/graphics/FontRanges.h:
(WebCore::FontRanges::isGeneric const):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/text/CharacterProperties.h:
(WebCore::isPrivateUseAreaCharacter):

Canonical link: https://commits.webkit.org/269524@main
